### PR TITLE
Reformat dataloader module

### LINKED ISSUE #17: Make generation of random crops fast
The current implementation of our [random crop](https://github.com/soumik12345/mirnetv2/blob/main/mirnetv2/dataloader/base/base_dataloader.py#L37) function uses slicing. We should be able to make a TensorFlow implementation of this operation to make it faster.

### LINKED ISSUE #18: Constraints the image type
https://github.com/soumik12345/mirnetv2/blob/0dd215b926f772f34c7ed208f0ac7fc24e7e4af4/mirnetv2/dataloader/base/base_dataloader.py#L33

I feel this link constraints the image type to only `pngs`. I would like to hear what you all think about this.

### PR DESCRIPTION
**Pull request by ariG23498**
Fixes #17 
Fixes #18

### CODE REVIEW

**Comment by ariG23498** [comment]
@soumik12345 I am still working on this. **Do not merge**. Just review as far as I have made the changes.

**Comment by soumik12345** [comment]
> @soumik12345 I am still working on this. **Do not merge**. Just review as far as I have made the changes.

While you're working on a PR, I recommend you mark it as draft. Mark it ready for review once you think its done.

File: restorers/dataloader/base/base_dataloader.py [hunk 1]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
+        Returns:
+            A tuple of random cropped image.
+        """
+        # Check whether the image size is smaller than the original image
+        image_size = tf.minimum(self.image_size, tf.shape(low_light_image)[0])
+
+        # Concatenate the low light and enhanced image
+        concatenated_image = tf.concat(
+            [low_light_image[None, ...], enhanced_image[None, ...]],
+            axis=0,  # Note the batch dims
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:76]: @ariG23498 Shouldn't the concatenation be on `axis=-1`?

File: restorers/dataloader/base/base_dataloader.py [hunk 2]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
-        input_image_cropped.set_shape([self.image_size, self.image_size, 3])
-        gt_image_cropped.set_shape([self.image_size, self.image_size, 3])
-        return input_image_cropped, gt_image_cropped
-
-    def resize(self, input_image, enhanced_image):
-        input_image = tf.image.resize(
-            input_image,
-            size=[self.image_size, self.image_size],
+        cropped_low_light_image, cropped_enhanced_image = tf.split(
+            cropped_concatenated_image, num_or_size_splits=2, axis=0
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:84]: Same here...

**Review by soumik12345 — COMMENTED** [review]
@ariG23498 I have a few queries at this point.

File: restorers/dataloader/base/base_dataloader.py (hunk 1 above)
Comment (ariG23498) [reply restorers/dataloader/base/base_dataloader.py:76]: My rationale behind this code is:
- Including a batch dimension with `[None, ...]`
- And concatenating the images from the batch.

This makes more intuitive sense of how we would do a random crop.

File: restorers/dataloader/base/base_dataloader.py (hunk 2 above)
Comment (ariG23498) [reply restorers/dataloader/base/base_dataloader.py:84]: To use the batch dim and then split it again I am using `axis=0`.

File: restorers/dataloader/base/base_dataloader.py [hunk 3]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
-        input_image_cropped = input_image[
-            crop_height : crop_height + self.image_size,
-            crop_width : crop_width + self.image_size,
-        ]
-        gt_image_cropped = gt_image[
-            crop_height : crop_height + self.image_size,
-            crop_width : crop_width + self.image_size,
-        ]
-        input_image_cropped.set_shape([self.image_size, self.image_size, 3])
-        gt_image_cropped.set_shape([self.image_size, self.image_size, 3])
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:54]: @ariG23498 I've been testing the changes and it seems like we need these lines, otherwise, the tensor specs of the dataset are gonna be `None`s.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@ariG23498 Please let me know your thoughts.

File: restorers/dataloader/base/base_dataloader.py [hunk 4]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
+        Mapping function for `tf.data.Dataset`. Loads the image from file path,
+        applies `random_crop` based on a boolean flag.
+
+        Args:
+            low_light_image_path (`str`): The file path for low light image.
+            enhanced_image_path (`str`): The file path for enhanced image.
+            apply_crop (`bool`): Boolean flag to condition random cropping.
+        """
+        # Read the image off the file path.
         low_light_image = self.read_image(low_light_image_path)
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:131]: Same here,
* `low_light_image` -> `input_image`
* `low_light_image_path` -> `input_image_path `

File: restorers/dataloader/base/base_dataloader.py [hunk 5]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
         enhanced_image = self.read_image(enhanced_image_path)
+
+        # Apply random cropping based on the boolean flag.
         low_light_image, enhanced_image = (
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:135]: Same here, `low_light_image` -> `input_image`

File: restorers/dataloader/base/base_dataloader.py [hunk 6]
```diff
@@ -28,46 +39,99 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
             self.random_crop(low_light_image, enhanced_image)
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:136]: Same here, `low_light_image` -> `input_image`

File: restorers/dataloader/base/base_dataloader.py [hunk 7]
```diff
@@ -77,20 +141,35 @@ def load_image(self, low_light_image_path, enhanced_image_path, apply_crop):
 
     def build_dataset(
         self,
-        low_light_images,
-        enhanced_images,
-        batch_size,
-        apply_crop,
-        apply_augmentations,
+        low_light_images: List[str],
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:144]: Same here, `low_light_images` -> `input_images`

File: restorers/dataloader/base/base_dataloader.py [hunk 8]
```diff
@@ -77,20 +141,35 @@ def load_image(self, low_light_image_path, enhanced_image_path, apply_crop):
...
+        enhanced_images: List[str],
+        batch_size: int,
+        apply_crop: bool,
+        apply_augmentations: bool,
     ):
+        """
+        Function to build the dataset.
+
+        Args:
+            low_light_images (`List[str]`): A list of image filenames.
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:154]: Same here, `low_light_images` -> `input_images`

File: restorers/dataloader/base/base_dataloader.py [hunk 9]
```diff
@@ -77,20 +141,35 @@ def load_image(self, low_light_image_path, enhanced_image_path, apply_crop):
...
+            enhanced_images (`List[str]`): A list of image filenames.
+            batch_size (`int`): Number of images in a single batch.
+            apply_crop (`bool`): Boolean flag to condition cropping.
+            apply_augmentations (`bool`): Boolean flag to condition augmentations.
+        """
+        # Build a `tf.data.Dataset` from the filenames.
         dataset = tf.data.Dataset.from_tensor_slices(
             (low_light_images, enhanced_images)
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:162]: Same here, `low_light_images` -> `input_images`

File: restorers/dataloader/base/base_dataloader.py [hunk 10]
```diff
@@ -103,18 +182,27 @@ def build_dataset(
...
+            batch_size (`int`): Number of images in a single batch.
+
+        Returns:
+            A tuple of `tf.data.Dataset` for training and validation.
+        """
         train_dataset = self.build_dataset(
-            self.train_low_light_images,
-            self.train_enhanced_images,
-            batch_size,
+            low_light_images=self.train_low_light_images,
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:196]: Same here,
* `low_light_images` -> `input_images`
* `train_low_light_images ` -> `train_input_images `

File: restorers/dataloader/base/base_dataloader.py [hunk 11]
```diff
@@ -103,18 +182,27 @@ def build_dataset(
...
+            enhanced_images=self.train_enhanced_images,
+            batch_size=batch_size,
             apply_crop=True,
             apply_augmentations=True,
         )
         val_dataset = self.build_dataset(
-            self.val_low_light_images,
-            self.val_enhanced_images,
-            batch_size,
+            low_light_images=self.val_low_light_images,
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:203]: Same here,
* `low_light_images` -> `input_images`
* `val_low_light_images` -> `val_input_images `

**Review by soumik12345 — COMMENTED** [review]
@ariG23498 Here's a minor suggestion to update the names of a few variables [restorers/dataloader/base/base_dataloader.py](https://github.com/soumik12345/mirnetv2/blob/fb311ad70dbbf8b1ca99b8073484905a41321e7f/restorers/dataloader/base/base_dataloader.py). Feel free to ignore it if you think its not necessary, otherwise you can just replace all the instances of `low_light` with `input` in the aforementioned file 😅

File: restorers/dataloader/base/base_dataloader.py (hunk 3 above)
Comment (soumik12345) [reply restorers/dataloader/base/base_dataloader.py:54]: Turns out that removing these lines was not the issue, rather, it was [this](https://github.com/soumik12345/mirnetv2/blob/ari/data/restorers/dataloader/base/base_dataloader.py#L53) line. Turns out, changing it to the following solves the issue
```python
image = tf.io.decode_image(image, channels=3, expand_animations=False)
```
Reference: https://www.tensorflow.org/api_docs/python/tf/io/decode_image

File: restorers/dataloader/base/base_dataloader.py [hunk 12]
```diff
@@ -28,69 +39,145 @@ def fetch_dataset(self, val_split: float, visualize_on_wandb: bool):
...
+
+        Returns:
+            A tuple of random cropped image.
+        """
+        # Check whether the image size is smaller than the original image
+        image_size = tf.minimum(self.image_size, tf.shape(input_image)[0])
+
+        # Concatenate the low light and enhanced image
+        concatenated_image = tf.concat(
+            [input_image[None, ...], enhanced_image[None, ...]],
```
Comment (soumik12345) [inline restorers/dataloader/base/base_dataloader.py:75]: Hi @ariG23498, as per https://github.com/soumik12345/mirnetv2/issues/17 we decided upon the removal of slice operations to make our data loading faster. I think we could avoid the slicing here entirely if we concatenate using the last axis instead of the first one. This is just a minor suggestion, feel free to ignore it.

**Review by soumik12345 — APPROVED** [review]
@ariG23498 I think that we're ready to merge.

File: restorers/dataloader/base/base_dataloader.py (hunk 12 above)
Comment (ariG23498) [reply restorers/dataloader/base/base_dataloader.py:75]: +1
